core/src/main/java/org/apache/spark/shuffle/api/SupportsTransferTo.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle.api;
+
+import java.io.IOException;
+
+import org.apache.spark.annotation.Experimental;
+
+/**
+ * :: Experimental ::
+ * Indicates that partition writers can transfer bytes directly from input byte channels to
+ * output channels that stream data to the underlying shuffle partition storage medium.
+ * <p>
+ * This API is separated out for advanced users because it only needs to be used for
+ * specific low-level optimizations. The idea is that the returned channel can transfer bytes
+ * from the input file channel out to the backing storage system without copying data into
+ * memory.
+ * <p>
+ * Most shuffle plugin implementations should use {@link ShufflePartitionWriter} instead.
+ *
+ * @since 3.0.0
+ */
+@Experimental
+public interface SupportsTransferTo extends ShufflePartitionWriter {
```

Comment:
We originally put it here so that anyone who looks at the interface - documentation, or just looking at the source code in general - could be isolated from this functionality entirely. Basically people don't have to worry about whether or not they have to implement this. But I'm not sure - is this too rigorous?

Reply:
Mostly I think this interface is weird the way it's currently defined.

The name is awkward. It sounds like the name of a mix-in interface (e.g. `extends ShufflePartitionWriter with SupportsTransferTo` in scala, except I'd call it `TransferToSupport` in that case). But Java can't enforce at compile time that the mix-in is applied to a `ShufflePartitionWriter`, so that's why you'd need inheritance.

But given that you can have default implementations in Java interfaces, is the extra interface really helping? The javadoc for the method with the default implementation will already make it clear that sub-classes don't necessarily need to worry about it, no?

Reply:
(Another option is to hide this completely inside Spark-private classes, so that the local version can implement it, but the optimization is not exposed outside of Spark.)

Reply:
I tend to agree with @vanzin , at least we could change to `TransferToSupport`. The mix-in semantics here is a little strange for Java, and actually we use it more like inheritance.